File: Lab/Lab6/Lab6.py
from collections import deque

def canFinish(num_courses, prerequisites):
    graph = [[] for _ in range(num_courses)]
    
    for prerequisite in prerequisites:
        if prerequisite[0] == prerequisite[1]:
            return False  # A course that points to itself
        graph[prerequisite[0]].append(prerequisite[1])
    
    q = deque()
    processed = 0
    
    while processed != num_courses:
        for i in range(len(graph)):
            if not graph[i]:
                q.append(i)
        
        if not q:
            return False  # Cycle detected
        
        while q:
            curr = q.popleft()
            graph[curr].append(num_courses + 1)
            processed += 1
            print(curr, processed)
            for i in range(len(graph)):
                if curr in graph[i]:
                    graph[i].remove(curr)
    
    return True

# Test
array = [
    [0, 1],
    [1, 0]
]

print(canFinish(3, array))

File: Lab/Lab6/Lab6.java
package Lab.Lab6;

import java.util.*;

public class Lab6{
    public static void main(String[] args) {
        int[][] array = {
            {0, 1},
            {1, 0}
        };
        
        
        System.out.println(canFinish(3, array));
    }
    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i < numCourses; i++){
            graph.add(i , new ArrayList<Integer>());
        }
        for(int i = 0; i < prerequisites.length; i++){
            if(prerequisites[i][0] == prerequisites[i][1]){
                return false; // A course that points to itself
            }
            graph.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }
        Queue<Integer> q = new LinkedList<Integer>();
        int processed = 0;
        while(processed != numCourses){
            for(int i = 0; i < graph.size(); i++){
                if(graph.get(i).size() == 0){
                    q.add(i);
                }
            }
            if(q.isEmpty()){
                return false; //cycle detected
            }
            while(!q.isEmpty()){
                Integer curr = q.poll();
                graph.get(curr).add(numCourses+1);
                processed++;
                System.out.println(curr + " " + processed);
                for(int i = 0; i < graph.size(); i++){
                    for(int j = 0; j < graph.get(i).size(); j++){
                        if(graph.get(i).get(j) == curr){
                            graph.get(i).remove(j);
                        }
                    }
                    
                }
            }
        }
        return true;
    }
}